### PR TITLE
Dynamic thread-safe hashmap

### PR DESCRIPTION
**Pull request by LytixDev**

### CODE REVIEW

**Comment by LytixDev** [comment]
very much still WIP!

**Comment by callumgran** [comment]
Instead of taking extra as an input, could we not just use the last byte of whatever the key is as this is likely to not be NULL, this also saves the user from having to send an extra key each time?

This could easily be done by doing something like this: `extra = ((key >> (key_size - 1) * 8) & 0xFF);`.

**Comment by LytixDev** [comment]
We could, however every string is terminated by the NULL byte and will therefore always have the last byte.

If we have a struct:
```c
struct my_struct {
    void *data;
    enum my_enum a;
};
```
If we look at the pure last byte, it will be whatever the value of `a` is. What if most structs tend to have the same value for `a`? Or at least little variance? Say the enum `my_enum` only has 4 possible values. We would get a lot of "false positives" (1/4). Since the hash function should add a lot of entropy, by storing one byte we only get a "false positive match" every 1/255 times.

memcmp() compares all the memory. It does not prematurely exit if the given parameters are not equal. That's why it's nice to skip the memcmp all together if possible. Instead of using memcmp() we could use BYTE_ASSERT and prematurely exit if one byte does not match. With this solution we may skip storing the extra byte all together.

**Comment by LytixDev** [comment]
Also, the extra byte should be taken from the hash before it is bit shifted to the right.
If our hashmap is less or equal to 2**8 (255) then taking the last byte after right shifting is useless! The hash *is one only byte* ...

File: hashmap.c [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        if (bucket->overflow == NULL) {
+            /* 
+             * No more space in direct pointers and no overflow bucket. A duplicate can therefore
+             * not exist and we can safely create an overflow bucket and insert the item.
+             */
+            bucket->overflow = overflow_alloc();
+            insert_into_overflow(bucket->overflow, key, key_size, value, extra);
+        } else {
+            insert_into_overflow(bucket->overflow, key, key_size, value, extra);
+        }
```
Comment (LytixDev) [inline hashmap.c:115]: i forgor

File: hashmap.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/* Written by Nicolai H. Brand 2023 */
+#ifndef HASHMAP_H
+#define HASHMAP_H
+
+//#define HASHMAP_THREAD_SAFE
+#define HASHMAP_BUCKET_SIZE 6
+#define HASHMAP_STARTING_OVERFLOW_BUCKET_SIZE 4
+#define HASHMAP_STARTING_BUCKETS_log2 3
+#define N_BUCKETS(log2) (2 << ((log2) - 1))
```
Comment (LytixDev) [inline hashmap.h:9]: bro just do 1 << log2

**Pull request merged**